File: src/endstone_endweave/codec/types/gameplay.py
"""Gameplay-related compound types shared across protocol handlers."""

from dataclasses import dataclass
from typing import Any

from ..reader import PacketReader
from ..writer import PacketWriter
from .primitives import (
    BOOL,
    FLOAT_LE,
    INT_LE,
    STRING,
    UINT_LE,
    UVAR_INT,
    VAR_INT,
    Type,
)

# ---------------------------------------------------------------------------
# Game rules (conditional union per entry)
# ---------------------------------------------------------------------------

_RULE_VALUE_TYPES: dict[int, Type[Any]] = {
    1: BOOL,
    2: VAR_INT,
    3: FLOAT_LE,
}


@dataclass
class GameRule:
    """A game rule entry with typed value.

    Attributes:
        name: Rule name string.
        editable: Whether the rule can be modified.
        type_id: Rule type (1=bool, 2=varint, 3=float).
        value: Rule value (type depends on type_id).
    """

    name: str
    editable: bool
    type_id: int
    value: Any


class _GameRulesType(Type[list["GameRule"]]):
    """UVAR_INT count + per-rule entries with conditional value type."""

    def read(self, reader: PacketReader) -> list[GameRule]:
        count = UVAR_INT.read(reader)
        rules: list[GameRule] = []
        for _ in range(count):
            name = STRING.read(reader)
            editable = BOOL.read(reader)
            type_id = UVAR_INT.read(reader)
            value_type = _RULE_VALUE_TYPES.get(type_id)
            if value_type is None:
                raise ValueError(f"Unknown game rule type: {type_id}")
            rules.append(GameRule(name=name, editable=editable, type_id=type_id, value=value_type.read(reader)))
        return rules

    def write(self, writer: PacketWriter, value: list[GameRule]) -> None:
        UVAR_INT.write(writer, len(value))
        for rule in value:
            STRING.write(writer, rule.name)
            BOOL.write(writer, rule.editable)
            UVAR_INT.write(writer, rule.type_id)
            value_type = _RULE_VALUE_TYPES.get(rule.type_id)
            if value_type is None:
                raise ValueError(f"Unknown game rule type: {rule.type_id}")
            value_type.write(writer, rule.value)


# ---------------------------------------------------------------------------
# Experiments (count + entries)
# ---------------------------------------------------------------------------


@dataclass
class Experiment:
    """An experiment entry (name + enabled flag).

    Attributes:
        name: Experiment name string.
        enabled: Whether the experiment is enabled.
    """

    name: str
    enabled: bool


class _ExperimentsType(Type[list[Experiment]]):
    """Count-prefixed list of (STRING + BOOL) experiment entries.

    Args:
        count_type: Type used for the list count (UINT_LE for v924+, INT_LE for v860/v898).
    """

    def __init__(self, count_type: Type[int] = UINT_LE) -> None:
        self._count_type = count_type

    def read(self, reader: PacketReader) -> list[Experiment]:
        count = self._count_type.read(reader)
        return [Experiment(name=STRING.read(reader), enabled=BOOL.read(reader)) for _ in range(count)]

    def write(self, writer: PacketWriter, value: list[Experiment]) -> None:
        self._count_type.write(writer, len(value))
        for exp in value:
            STRING.write(writer, exp.name)
            BOOL.write(writer, exp.enabled)


# ---------------------------------------------------------------------------
# Singletons
# ---------------------------------------------------------------------------

GAME_RULES = _GameRulesType()
EXPERIMENTS = _ExperimentsType(UINT_LE)
EXPERIMENTS_V860 = _ExperimentsType(INT_LE)

File: src/codec/types/gameplay.cpp
/// GameRule, Experiment codec implementation.

#include "endweave/codec/types/gameplay.h"
#include "endweave/codec/types/primitives.h"

namespace endweave {

template <> auto PacketReader::read<game_rules>() -> std::expected<std::vector<GameRule>, ReadError> {
    auto count = read_uvarint();
    if (!count) return std::unexpected(count.error());
    std::vector<GameRule> rules;
    rules.reserve(*count);
    for (std::uint32_t i = 0; i < *count; ++i) {
        GameRule r;
        auto name = read_string();
        if (!name) return std::unexpected(name.error());
        r.name = std::move(*name);
        auto editable = read_bool();
        if (!editable) return std::unexpected(editable.error());
        r.editable = *editable;
        auto type_id = read_uvarint();
        if (!type_id) return std::unexpected(type_id.error());
        r.type_id = *type_id;
        switch (*type_id) {
        case 1: { auto v = read_bool(); if (!v) return std::unexpected(v.error()); r.value = *v; break; }
        case 2: { auto v = read_varint(); if (!v) return std::unexpected(v.error()); r.value = *v; break; }
        case 3: { auto v = read_float_le(); if (!v) return std::unexpected(v.error()); r.value = *v; break; }
        default: return std::unexpected(ReadError::out_of_bounds);
        }
        rules.push_back(std::move(r));
    }
    return rules;
}

template <> void PacketWriter::write<game_rules>(const std::vector<GameRule>& val) {
    write_uvarint(static_cast<std::uint32_t>(val.size()));
    for (auto& r : val) {
        write_string(r.name);
        write_bool(r.editable);
        write_uvarint(r.type_id);
        switch (r.type_id) {
        case 1: write_bool(std::get<bool>(r.value)); break;
        case 2: write_varint(std::get<std::int32_t>(r.value)); break;
        case 3: write_float_le(std::get<float>(r.value)); break;
        default: break;
        }
    }
}

namespace {
template <typename CountTag>
std::expected<std::vector<Experiment>, ReadError> read_experiments_impl(PacketReader& reader) {
    auto count_result = reader.read<CountTag>();
    if (!count_result) return std::unexpected(count_result.error());
    auto count = static_cast<std::uint32_t>(*count_result);
    std::vector<Experiment> exps;
    exps.reserve(count);
    for (std::uint32_t i = 0; i < count; ++i) {
        Experiment e;
        auto name = reader.read_string();
        if (!name) return std::unexpected(name.error());
        e.name = std::move(*name);
        auto enabled = reader.read_bool();
        if (!enabled) return std::unexpected(enabled.error());
        e.enabled = *enabled;
        exps.push_back(std::move(e));
    }
    return exps;
}

template <typename CountTag>
void write_experiments_impl(PacketWriter& writer, const std::vector<Experiment>& val) {
    writer.write<CountTag>(static_cast<typename CountTag::value_type>(val.size()));
    for (auto& e : val) {
        writer.write_string(e.name);
        writer.write_bool(e.enabled);
    }
}
} // namespace

template <> auto PacketReader::read<experiments>() -> std::expected<std::vector<Experiment>, ReadError> {
    return read_experiments_impl<uint_le>(*this);
}
template <> void PacketWriter::write<experiments>(const std::vector<Experiment>& val) {
    write_experiments_impl<uint_le>(*this, val);
}
template <> auto PacketReader::read<experiments_v860>() -> std::expected<std::vector<Experiment>, ReadError> {
    return read_experiments_impl<int_le>(*this);
}
template <> void PacketWriter::write<experiments_v860>(const std::vector<Experiment>& val) {
    write_experiments_impl<int_le>(*this, val);
}

} // namespace endweave
